File: scripts/nucleo/Sensor.js
/**
 * Classe base para todos os sensores.
 * Subclasses (SensorEnergia, SensorTemperatura, SensorPorta) implementam
 * o cálculo de indicadores e a renderização específica.
 */
class Sensor {
  constructor({ id, label, type, group, status, fields = [] } = {}) {
    this.id = id;
    this.rotulo = label;
    this.tipo = type;
    this.grupo = group;
    this.status = status;
    this.campos = fields;
  }

  get ativo()      { return this.status === "ativo"; }
  get historico()  { return this.status === "historico"; }

  /** Cor temática do sensor (referência à paleta). */
  get cor() {
    return {
      energia:     "var(--azul-medio)",
      temperatura: "var(--ciano)",
      porta:       "var(--azul-claro)",
    }[this.tipo] ?? "var(--azul-profundo)";
  }

  /**
   * Devolve uma lista de indicadores calculados a partir dos pontos brutos.
   * Cada item: { rotulo, valor, sub?, severidade? }
   * Subclasse deve sobrescrever.
   */
  calcularIndicadores(/* pontos */) {
    return [];
  }

  /** Útil pra logs / debug. */
  toString() {
    return `[${this.tipo}] ${this.rotulo} (${this.id})`;
  }

  /** Helper: média numérica ignorando undefined. */
  static media(arr) {
    if (!arr.length) return 0;
    return arr.reduce((s, x) => s + (Number.isFinite(+x) ? +x : 0), 0) / arr.length;
  }
}

if (typeof window !== "undefined") window.Sensor = Sensor;

File: app_flutter/lib/data/models/sensor.dart
/// Modelo de sensor (espelha o JSON devolvido por `listar_catalogo`).
class Sensor {
  const Sensor({
    required this.id,
    required this.rotulo,
    required this.tipo,
    required this.grupo,
    required this.status,
    this.parametros = const {},
    this.personalidade,
  });

  final String id;
  final String rotulo;
  final String tipo;       // "energia" | "temperatura" | "porta"
  final String grupo;      // id do grupo (extrusao, camara_congelados, …)
  final String status;     // "ativo" | "historico"
  final Map<String, dynamic> parametros;
  final String? personalidade;

  bool get ativo     => status == 'ativo';
  bool get historico => status == 'historico';

  factory Sensor.fromJson(Map<String, dynamic> j) => Sensor(
        id: j['id'] as String,
        rotulo: (j['label'] ?? j['rotulo'] ?? j['id']) as String,
        tipo: (j['type'] ?? j['tipo'] ?? 'energia') as String,
        grupo: (j['group'] ?? j['grupo_id'] ?? j['grupo'] ?? '') as String,
        status: (j['status'] ?? 'ativo') as String,
        parametros: (j['parametros'] as Map?)?.cast<String, dynamic>() ?? const {},
        personalidade: j['personalidade'] as String?,
      );
}
